### PR TITLE
Fix missing correlation ID in log messages

### PR DESCRIPTION
**Pull request by irinaepshteyn**
Added distributed log tracing.

### CODE REVIEW

File: server/src/main/java/org/cloudfoundry/identity/uaa/web/UaaLogFilter.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        }
+    }
+
+    private void setCorrelationId(final HttpServletRequest request, final HttpServletResponse response) {
+        String correlationId = request.getHeader(CORRELATION_HEADER_NAME);
+        if (StringUtils.isEmpty(correlationId)) {
+            correlationId = UUID.randomUUID().toString();
+        }
+        MDC.put(CORRELATION_KEY_NAME, correlationId);
+        response.setHeader(CORRELATION_HEADER_NAME, correlationId);
```
Comment (metacubed) [inline server/src/main/java/org/cloudfoundry/identity/uaa/web/UaaLogFilter.java:47]: Do we need to add a response header? It is not part of the tracing standard, just something custom we used to do.

File: server/src/main/java/org/cloudfoundry/identity/uaa/web/UaaLogFilter.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    protected void doFilterInternal(final HttpServletRequest request,
+                                    final HttpServletResponse response, final FilterChain filterChain)
+        throws ServletException, IOException {
+        try {
+            setCorrelationId(request, response);
+            super.doFilterInternal(request, response, filterChain);
+        } finally {
+            clearMDC();
```
Comment (metacubed) [inline server/src/main/java/org/cloudfoundry/identity/uaa/web/UaaLogFilter.java:37]: (minor) Rename this to `clearCorrelationId` to match `setCorrelationId `?

File: server/src/main/java/org/cloudfoundry/identity/uaa/web/UaaLogFilter.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        }
+    }
+
+    private void setCorrelationId(final HttpServletRequest request, final HttpServletResponse response) {
+        String correlationId = request.getHeader(CORRELATION_HEADER_NAME);
+        if (StringUtils.isEmpty(correlationId)) {
+            correlationId = UUID.randomUUID().toString();
```
Comment (metacubed) [inline server/src/main/java/org/cloudfoundry/identity/uaa/web/UaaLogFilter.java:44]: This does not match the expected trace-id format. Check the trace-ids created by CF/AVI.

File: server/src/main/java/org/cloudfoundry/identity/uaa/web/UaaLogFilter.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class UaaLogFilter extends LogFilter {
+    static final String CORRELATION_HEADER_NAME = "X-B3-TraceId";
+    static final String CORRELATION_KEY_NAME = "traceId";
+
+    public UaaLogFilter(final LinkedHashSet<String> hostnames, final LinkedHashSet<String> zoneHeaders,
+                     final String defaultZone) {
```
Comment (metacubed) [inline server/src/main/java/org/cloudfoundry/identity/uaa/web/UaaLogFilter.java:21]: (minor) Whitespace

**Review by metacubed — CHANGES_REQUESTED** [review]

**Comment by akarollil** [comment]
Pushkaraj gave me background info for this. I don't understand it enough :), but guess after upgrade to 75.0.0, we tried removing code (https://github.com/predix/spring-log-filter/commit/6087cfc668d3e2b7d8a25d74ded82cfff3ff2693) that added the correlation ID hoping that Spring Sleuth (added/enabled in 75.0.0?) will take care of it for us? But it doesn't in some cases because UAA isn't fully spring-booted :D. So this commit is bringing in that customization but by extending spring-log-filter?

I had some comments about the commit message. I tend to follow guidelines mentioned in https://chris.beams.io/posts/git-commit/. So if we follow that, the commit message could be something like what is below. Subject line limited to 50 characters. A blank link separating subject and body. Body lines wrapped at 72. Title/subject should ideally be about what the commit achieves maybe? We used to have distributed log tracing, but it went away after upgrade changes in some cases?

```
Fix missing correlation ID in log messages

Add distributed log tracing. In addition to Predix Spring Log Filter, it
enables UAA to populate the corr field of the log using the 'traceId'
attribute of the SLF4J MDC: predix/spring-log-filter@49d6ecd.

Note, spring-cloud-sleuth (Reference:
https://cloud.spring.io/spring-cloud-sleuth) doesn't seem to do this as
expected. Since UAA is "spring boot wannabe", sleuth doesn't correctly
instrument UAA beans for distributed tracing. Spring-cloud-sleuth is
boot based and only works with spring boot applications.

```

**Comment by irinaepshteyn** [comment]
@metacubed, @akarollil, implemented requested changes.  Please review again.

**Comment by akarollil** [comment]
Looks good to me, I will let Pushkaraj approve.

**Review by metacubed — APPROVED** [review]

**Pull request merged**